include/mbgl/style/conversion/data_driven_property_value.hpp

```diff
@@ -47,13 +47,15 @@ struct Converter<DataDrivenPropertyValue<T>> {
             return {};
         } else if (!(*expression).isFeatureConstant() || !(*expression).isZoomConstant()) {
             return { std::move(*expression) };
-        } else {
+        } else if ((*expression).getExpression().getKind() == Kind::Literal) {
             optional<T> constant = fromExpressionValue<T>(
-                dynamic_cast<const Literal&>((*expression).getExpression()).getValue());
+                static_cast<const Literal&>((*expression).getExpression()).getValue());
             if (!constant) {
                 return {};
             }
             return DataDrivenPropertyValue<T>(*constant);
+        } else {
+            return {};
```

Comment:
If we're going to do defensive programming for the case where an expected literal isn't a literal, there should be an `assert(false)` here and something like `error = { "expected a literal expression" }`. (x2)